### ISSUE TITLE
OIDC & Custom Nonce/Correlation

### ISSUE THREAD

**Issue by justinhelgerson** [aspnet/Security#1803]
I'm using the Okta [sign-in library](https://developer.okta.com/code/javascript/okta_sign-in_widget) to add a sign in experience to my application.  I have it configured to route the response to the default `/signin-oidc` endpoint used by the middleware (an HTTP POST including a `code`, `id_token` and `state` in the body of the response).  However, this results in the following error:

> System.Exception: Correlation failed.
>    at Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler`1.<HandleRequestAsync>d__12.MoveNext()

From poking around the source code of the middleware, I suspect this is because the cookies being searched aren't available (because the request to initiate the process isn't happening by the middleware, it's happening from the Okta library).  The library seems to create two cookies: `okta-oauth-nonce` and `okta-oauth-state`.

In short, is it possible to tell the OIDC middleware to validate based on these cookies instead of the defaults?

**Comment by Tratcher** [comment]
No. You could change the names but it wouldn't be able to parse the state parameter.

If the Okta library is generating the challenge then it should handle the response. Can it?

**Comment by justinhelgerson** [comment]
Out of curiosity, is it because those values are wrapped around the aspnet data protection API so they wouldn't validate?

You're absolutely correct that I can get the response with the library.  I've got access to a JSON response with an `id_token` and `claims`.  However, I want to leverage the security API's of .NET (claims, authorize attributes, policies, etc).  My app isn't a single page application, so keeping that data on the client isn't very useful.

I suppose I could submit the JWT `id_token` to my app, validate it and finally manually auth the user and add claims.  It's a little more work than I'd like, but that could be an option.

Do you have any other suggestions that I might be overlooking?

**Comment by justinhelgerson** [comment]
Isn't the JwtBearer more geared towards a single page app or API flow?  Where the client would hold onto the token and present it with each request?

I'm essentially looking for the flow provided by the `AddOpenIdConnect` extension where a cookie is set including the claims inside the token.  In this scenario, I'll be losing access to the token on the next page.

**Comment by Tratcher** [comment]
The Okta plugin you linked to is all client side, that matches the SPA scenario. If you want to transition to a normal cookie flow you could use JwtBearer as the intermediary and then call SignIn via CookieAuth.

**Comment by brockallen** [comment]
> The Okta plugin you linked to is all client side, that matches the SPA scenario. If you want to transition to a normal cookie flow you could use JwtBearer as the intermediary and then call SignIn via CookieAuth.

Make sure you're validating the token really well in that workflow -- if you only validate that it's from okta then possibly any other access token from okta could be presented t your api to issue a cookie. I'd think you also need to check the client_id to ensure it's an access token from your SPA. That's not a normal part of access token validation. The devil's in the details, of course, and I don't know what tokens from okta look like (what issuer, what signing creds, and how that works in their multi-tenancy offering, etc).
